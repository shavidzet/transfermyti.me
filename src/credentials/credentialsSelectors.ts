import { createSelector } from "reselect";
import { toolNameByMappingSelector } from "~/allEntities/allEntitiesSelectors";
import {
  CredentialsByMappingModel,
  CredentialsModel,
  FetchStatus,
  Mapping,
  ReduxState,
  ToolName,
  ValidationErrorsByMappingModel,
} from "~/typeDefs";

export const validationFetchStatusSelector = (state: ReduxState): FetchStatus =>
  state.credentials.validationFetchStatus;

export const validationErrorsByMappingSelector = createSelector(
  (state: ReduxState) => state.credentials.validationErrorsByMapping,
  (validationErrorsByMapping): ValidationErrorsByMappingModel =>
    validationErrorsByMapping,
);

export const hasValidationErrorsSelector = createSelector(
  validationErrorsByMappingSelector,
  (validationErrorsByMapping): boolean =>
    Object.values(validationErrorsByMapping).some(Boolean),
);

const sourceCredentialsSelector = createSelector(
  (state: ReduxState) => state.credentials.source,
  sourceCredentials => sourceCredentials,
);

const targetCredentialsSelector = createSelector(
  (state: ReduxState) => state.credentials.target,
  targetCredentials => targetCredentials,
);

export const credentialsByMappingSelector = createSelector(
  sourceCredentialsSelector,
  targetCredentialsSelector,
  (sourceCredentials, targetCredentials): CredentialsByMappingModel => ({
    [Mapping.Source]: sourceCredentials,
    [Mapping.Target]: targetCredentials,
  }),
);

export const credentialsByToolNameSelector = createSelector(
  toolNameByMappingSelector,
  sourceCredentialsSelector,
  targetCredentialsSelector,
  (
    toolNameByMapping,
    sourceCredentials,
    targetCredentials,
  ): Record<ToolName, CredentialsModel> =>
    ({
      [toolNameByMapping[Mapping.Source]]: sourceCredentials,
      [toolNameByMapping[Mapping.Target]]: targetCredentials,
    } as Record<ToolName, CredentialsModel>),
);
