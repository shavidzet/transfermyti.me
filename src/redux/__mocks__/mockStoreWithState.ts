import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
