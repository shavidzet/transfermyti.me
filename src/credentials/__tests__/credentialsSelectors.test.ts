import cases from "jest-in-case";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import * as credentialsSelectors from "../credentialsSelectors";

const TEST_STATE = {
  ...state,
  credentials: {
    ...state.credentials,
    source: {
      apiKey: "CLOCKIFYAPI",
      email: "clockify@test.com",
      userId: "abc",
    },
    target: {
      apiKey: "TOGGLAPI",
      email: "toggl@test.com",
      userId: "123",
    },
  },
};

describe("within credentialsSelectors", () => {
  cases(
    "selectors that directly access state return the correct value",
    options => {
      const result = options.selector(state);

      expect(result).toEqual(options.expected);
    },
    [
      {
        name:
          "validationFetchStatusSelector returns state.validationFetchStatus",
        selector: credentialsSelectors.validationFetchStatusSelector,
        expected: TEST_STATE.credentials.validationFetchStatus,
      },
      {
        name:
          "validationErrorsByMappingSelector returns state.validationErrorsByMapping",
        selector: credentialsSelectors.validationErrorsByMappingSelector,
        expected: TEST_STATE.credentials.validationErrorsByMapping,
      },
    ],
  );

  test("the hasValidationErrorsSelector returns true if validation errors present in state", () => {
    const result = credentialsSelectors.hasValidationErrorsSelector(TEST_STATE);

    expect(result).toBe(true);
  });

  cases(
    "the selectors match their snapshots",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the credentialsByMappingSelector",
        selector: credentialsSelectors.credentialsByMappingSelector,
      },
      {
        name: "for the credentialsByToolNameSelector",
        selector: credentialsSelectors.credentialsByToolNameSelector,
      },
    ],
  );
});
