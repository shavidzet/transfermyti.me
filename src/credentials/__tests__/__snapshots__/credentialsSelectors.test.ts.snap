// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within credentialsSelectors the selectors match their snapshots for the credentialsByMappingSelector 1`] = `
Object {
  "source": Object {
    "apiKey": "CLOCKIFYAPI",
    "email": "clockify@test.com",
    "userId": "abc",
  },
  "target": Object {
    "apiKey": "TOGGLAPI",
    "email": "toggl@test.com",
    "userId": "123",
  },
}
`;

exports[`within credentialsSelectors the selectors match their snapshots for the credentialsByToolNameSelector 1`] = `
Object {
  "clockify": Object {
    "apiKey": "TOGGLAPI",
    "email": "toggl@test.com",
    "userId": "123",
  },
  "toggl": Object {
    "apiKey": "CLOCKIFYAPI",
    "email": "clockify@test.com",
    "userId": "abc",
  },
}
`;
